### PR TITLE
feat: params in header for profile generate and assemble

### LINKED ISSUE #923: As an author of profiles I would like to use the markdown documents to set parameters
## Issue description / feature objectives
As an author of profiles I would like to use the markdown documents to set parameters
As a manager of profile projects I would like to control which profiles can set parameters.

The operational observation is that profiles provide the perfect vehicle to set parameters. Parameters are typically set such they are organisationally consistent (for example frequency of pen-testing is an organisational edict and should not vary system to system).

Given this we want to allow users to set parameters explicilty in the profiles. By this I mean CLI args control whether or not parameters can be set.

Behaviour:
`trestle author profile-generate` and `trestle author profile-assemble` get an additional `--set-params` argument.

## profile generate:
If `--set-params` is passed a header is generated where:
```yaml
---
x-trestle-set-params:
   param_1_id:
   param_2_id: 
```
is generated per control. A RHS is set if and only if the *CURRENT* profile has a `set-parameter` for that parameter. e.g. if `param_1_id` is set to foo in the current profile it would generate:
```yaml
---
x-trestle-set-params:
   param_1_id: 'foo'
   param_2_id: 
```
The existing parameter substitution would occur as-is into the prose.

## profile assemble:
if `--set-params` is passed (and the `x-trestle-set-params` header exists): The set-params dictionary in the profile is rebuilt based on all the parameters that have parameters set in the markdown documents.

## Caveats / Assumptions
- As similar workflow *could* be applied to ssp_generate and ssp_assemble, however, we will not do so to provide SoD unless raised by trestle users.

### LINKED ISSUE #933: user-entered markdown is sometimes overridden when running `ssp-generate`
## Describe the bug

In the [ssp generate documentation](https://ibm.github.io/compliance-trestle/tutorials/ssp_profile_catalog_authoring/ssp_profile_catalog_authoring/#trestle-author-ssp-generate-and-trestle-author-ssp-assemble) we see this paragraph:

```
If you edit the control markdown files you may run ssp-generate again and your edits will not be overwritten. When writing out the markdown for a control, any existing markdown for that control will be read and the response text for each part will be re-inserted into the new markdown file. If the new markdown has added parts the original responses will be placed correctly in the new file, but if any part is removed from the control in the update then any corresponding prose will be lost.
```

This is only true, I think, if the control has multiple parts to it. If not, then the file is re-generated and any previous edits are lost. And looking at the code, `existing_text` is only used when control parts are present, which seems to support my hypothesis. This [comment](https://github.com/IBM/compliance-trestle/blob/develop/trestle/core/control_io.py#L220-L223) in the relevant bit of the code seems to suggest that this once worked with top level elements. 

## To Reproduce

1. Create a new profile

```json
{
  "profile": {
    "uuid": "b7ac2bb6-75ec-411e-a162-378f50ac4ea3",
    "metadata": {
      "title": "Sample profile.",
      "last-modified": "2021-11-17T15:46:14.402466-08:00",
      "version": "0.0.0",
      "oscal-version": "1.0.0"
    },
    "imports": [
      {
        "href": "https://github.com/usnistgov/oscal-content/raw/master/nist.gov/SP800-53/rev5/json/NIST_SP-800-53_rev5_MODERATE-baseline-resolved-profile_catalog.json",
        "include-controls": [
          {
            "with-ids": [
              "ac-12"
            ]
          }
        ]
      }
    ]
  }
}
```

2. Run `trestle author ssp-generate -p test -o dist/system-security-plans/test` and observe that `ac-12.md` is generated.

3. Edit `ac-12.md`. 

```
sed -i "s/Add control/Something else/" dist/system-security-plans/test/ac-12.md
```

4. Re-run `trestle author ssp-generate -p test -o dist/system-security-plans/test`. The changes introduced above will be discarded. 

## Expected behavior

I would not want my control content to be overwritten when re-generating a ssp from a profile. Alternatively (the less better option) the documentation could be updated. 

## Screenshots / Logs.

Sample run

```
$ git diff | cat
diff --git a/dist/system-security-plans/test/ac-12.md b/dist/system-security-plans/test/ac-12.md
index 2aba1ba5..dbc5439f 100644
--- a/dist/system-security-plans/test/ac-12.md
+++ b/dist/system-security-plans/test/ac-12.md
@@ -22,6 +22,6 @@ ______________________________________________________________________

 ## What is the solution and how is it implemented?

-Add control implementation description here for control ac-12
+The solution is as follows...

 ______________________________________________________________________
$ trestle author ssp-generate -p test -o dist/system-security-plans/test
$ git diff | cat
<empty line>
```

## Environment

- OS: OSX Catalina 10.15.7
- Python version: 3.9.6
- Installed packages: compliance-trestle v0.30.0

### PR DESCRIPTION
**Pull request by fsuits**
## Types of changes

- [ ] Hot fix (emergency fix and release)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation (change which affects the documentation site)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Release (`develop` -> `main`)

## Quality assurance (all should be covered).

- [x] My code follows the code style of this project.
- [ ] Documentation for my change is up to date?
- [x] My PR meets testing requirements.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

## Summary

allow presenting and editing of parameters in markdown during profile-generate and assemble

closes #933 
closes #923 

## Key links:
Profile generate and assemble will handle a special section for parameter values in the yaml header
- [Sonar coverage](https://sonarcloud.io/dashboard?id=compliance-trestle)

# Before you merge

- Ensure it is a 'squash commit' if not a release.
- Ensure CI is currently passing
- Check sonar. If you are working for a fork a maintainer will reach out, if required.

### CODE REVIEW

File: trestle/core/control_io.py [hunk 1]
```diff
@@ -1048,7 +1089,11 @@ def read_new_alters(control_path: pathlib.Path) -> List[prof.Alter]:
                     new_alters.append(alter)
             else:
                 ii += 1
-        return new_alters
+        param_content = header.get(const.SET_PARAMS_TAG, {})
```
Comment (butler54) [inline trestle/core/control_io.py:1092]: I believe this needs to default to none.

Comment (butler54) [reply trestle/core/control_io.py:1092]: When being written out.

**Review by butler54 — CHANGES_REQUESTED** [review]
Please see any comments directly in the code below, however, here are some observed changes:

When set-parameters is used, and no control is provided, a strange empty dict exists if there are no parameters in the control e.g.:

```yaml
---
x-trestle-set-params: {}
---
```
if we could remove this it would be fantastic.

Currently `profile-assemble` will not remove a set-parameter if you zero it in the markdown.

E.g.
If I had:

```yaml
---
x-trestle-set-parameter:
     my-special-id: 'with a fancy value' 
---
```
Then ran profile assemble (followed by generate). We end up with the same content as expected. If I then zero the content e.g.

```yaml
---
x-trestle-set-parameter:
     my-special-id: '' 
---
```

and run profile-assemble then generate the zeroing does not stick (e.g. it reverts to ` my-special-id: 'with a fancy value' `).

**Comment by butler54** [comment]
Adding extra comments to our online discussion. The fixes for handling single response ssp sections appear to be a little broken as the part ID's are also incorrect. Improved handling within ssp-assemble was also recommended

**Comment by butler54** [comment]
Also - it looks like you are missing any `closes` tags in your PR.

File: trestle/core/control_io.py (hunk 1 above)
Comment (fsuits) [reply trestle/core/control_io.py]: I believe these are all addressed but I can't replicate the override of a parameter that was set to ''.  Please confirm and describe the test.

**Review by butler54 — APPROVED** [review]

**Comment by butler54** [comment]
I know there is a chunk of documentation to go. Please update the PR with the appropriate closes before squashing and merging.

**Pull request merged**